File: puzzle12/puzzle12.R
# PUZZLE 12A

# Read the puzzle input
x <- readLines("puzzle12.txt")
x <- strsplit(x, " ")
x <- lapply(x, function(x) list(strsplit(x[1], "")[[1]], as.integer(strsplit(x[2], ",")[[1]])))

# Looking at how many unknowns are in each record
table(sapply(x, function(x) sum(x[[1]] == "?")))
# Even the maximum of 2^19 (524288) is still doable

# A function that takes a given record and returns the number of possible arrangements
n_possible <- function(x) {
	# Creating a data.frame with all possible combinations
	comb_list <- list()
	for(i in 1:sum(x[[1]] == "?")) comb_list[[i]] <- c(0, 1)
	combinations <- as.matrix(expand.grid(comb_list))

	# We also get the indices of the unknown values
	indices <- which(x[[1]] == "?")

	# Initializing the output value to 0
	output <- 0

	# Then we loop over all combinations
	for(i in 1:nrow(combinations)) {
		x_copy <- x[[1]]
		# The ones are converted to #
		x_copy[indices[combinations[i, ] == 1]] <- "#"
		# The remaining values are converted to "."
		x_copy[x_copy == "?"] <- "."
		# Getting the run length encoding of the resulting values
		x_rle <- rle(x_copy)
		# We check if it matches the pattern
		output <- output + identical(x_rle$lengths[x_rle$values == "#"], x[[2]])
	}

	output
}

n_possibilities <- parallel::mclapply(1:length(x), function(i, x) {
			   message("Processing record ", i, " out of ", length(x))
			   output <- n_possible(x[[i]])
			   output
}, x = x, mc.cores = 8)

cat("Puzzle 12A:", sum(unlist(n_possibilities)), "\n")

# PUZZLE 12B

# A function that preprocesses a sequence to determine the maximum number of hashes
# that can fit for each index
preprocess_sequence <- function(x) {
	output <- numeric(length(x))

	for(i in 1:length(x)) {
		j <- i
		while(x[j] != "." && j <= length(x)) {
			output[i] <- output[i] + 1
			j <- j + 1
		}
	}

	output
}

# A function that returns a logical vector of positions of the hashes in a sequence
# The length is 1 more than the length of the sequence because we need to be able
# to tell whether there is a hash after a sequence
hash_positions <- function(x) {
	c(x == "#", FALSE)
}

# A function that computes the number of combinations to test
# given a number of '#' symbols to reach, the number already there,
# and the number of '?' symbols

# A recursive function that tests a target series of '#' symbols against a sequence
test_sequence <- function(sequence, seqlength, pp_vect, hpos, which_h, index, series) {

	# Initialize the return value to 0
	return_value <- 0

	# If this is the last element of the sequence then we can increment the return value
	# by the number of times that the series sites into the remaining sequence
	# But we must make sure that the admissible sequence is just one stretch
	if(length(series) == 1 && !any(hpos[(index + series):length(hpos)])) {
		# We meet all conditions, so we have found a possible match
		return(1)
	} else {
		# The series number eats up as many elements as the number + 1 (for the dot)
		if(!hpos[index + series[1]]) {
			# We need to find the next potential indices based on the series
			# size and the preprocessed sequence
			next_indices <- which(pp_vect >= series[2])
			next_indices <- next_indices[next_indices >= (index + series[1] + 1)]
			remaining_hashes <- which_h[which_h >= (index + series[1] + 1)]
			next_indices <- next_indices[next_indices <= suppressWarnings(min(remaining_hashes))]

			for(i in next_indices) {
				return_value <- return_value + test_sequence(sequence, seqlength, pp_vect, hpos,
									     which_h,
									     index = i,
									     series[-1])
			}
		}
	}

	return_value
}

prepare_indices <- function() {

}

# A function that determines the indices to test and adds them all to get the result
test_sequence_wrapper <- function(sequence, seqlength, pp_vect, hpos, index, series) {
	result <- 0

	which_h <- which(hpos)
	indices_to_test <- which(pp_vect >= series[1])
	indices_to_test <- indices_to_test[indices_to_test <= min(which_h)]

	for(i in indices_to_test) {
		result <- result + test_sequence(sequence, seqlength, pp_vect, hpos, which_h, i, series)
	}

	result
}

sum(sapply(x, function(x) test_sequence_wrapper(sequence = x[[1]],
						seqlength = length(x[[1]]),
						pp_vect = preprocess_sequence(x[[1]]),
						hpos = hash_positions(x[[1]]),
						index = 1,
						series = x[[2]])))

library(microbenchmark)
microbenchmark(sum(sapply(x, function(x) test_sequence_wrapper(sequence = x[[1]],
							       seqlength = length(x[[1]]),
							       pp_vect = preprocess_sequence(x[[1]]),
							       hpos = hash_positions(x[[1]]),
							       index = 1,
							       series = x[[2]]))),
	       times = 50)

test_sequence_wrapper(sequence = x[[1]][[1]],
		      seqlength = length(x[[1]][[1]]),
		      pp_vect = preprocess_sequence(x[[1]][[1]]),
		      hpos = hash_positions(x[[1]][[1]]),
		      index = 1,
		      series = x[[1]][[2]])

# Applying our new function to the modified input
n_possibilities <- parallel::mclapply(1:length(x), function(i, x) {
			   message("Processing record ", i, " out of ", length(x))
			   output <- test_sequence_wrapper(sequence = x[[i]][[1]],
							   seqlength = length(x[[i]][[1]]),
							   pp_vect = preprocess_sequence(x[[i]][[1]]),
							   hpos = hash_positions(x[[i]][[1]]),
							   index = 1,
							   series = x[[i]][[2]])
			   output
	       }, x = x, mc.cores = 8)

# Producing the duplicated records
x <- lapply(x, function(x) {
	       x[[1]] <- rep(c(x[[1]], "?"), 5)
	       x[[1]] <- x[[1]][-length(x[[1]])]
	       x[[2]] <- rep(x[[2]], 5)
	       x
})

n_possibilities <- parallel::mclapply(1:length(x), function(i, x) {
			   message("Processing record ", i, " out of ", length(x))
			   output <- test_sequence(sequence = x[[i]][[1]],
						   seqlength = length(x[[i]][[1]]),
						   pp_vect = preprocess_sequence(x[[i]][[1]]),
						   hpos = hash_positions(x[[i]][[1]]),
						   index = 1,
						   series = x[[i]][[2]])
			   output
	       }, x = x, mc.cores = 8)

cat("Puzzle 12B:", sum(unlist(n_possibilities)), "\n")

File: puzzle12/puzzle12.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

#define BUFSIZE 1000

using namespace std;

// Function prototypes
vector<int> preprocess_sequence(const string& sequence);
vector<bool> find_hashes(const string& sequence);
unsigned long long int test_sequence(const string& sequence, const vector<int>& series);
unsigned long long int seq_recurse(const string& sequence, const vector<int>& pp_vect, const vector<bool>& h_vect, int index,
		const vector<int>& series, int series_index, vector<vector<unsigned long long int>>& cache_num, vector<vector<bool>>& cache_bool);

int main() {

	// Create a string to hold the line
	ifstream input_file ("puzzle12.txt");

	string buffer, sequence, numbers;
	size_t next_pos = 0;
	unsigned long long int result = 0, solution_a = 0, solution_b = 0;

	// A vector of string that stores the sequences
	vector<string> sequences;

	// A vector of vectors that stores the series lengths
	vector<vector<int>> series;

	// Input loop
	while(getline(input_file, buffer)) {
		// Extracting the sequence of records
		sequence = buffer.substr(0, buffer.find(' '));
		sequences.push_back(sequence);

		// Extracting the comma-separated numbers and adding them to the vector
		numbers = buffer.substr(buffer.find(' ') + 1, string::npos);
		vector<int> i_vect;

		while((next_pos = numbers.find(',')) != string::npos) {
			i_vect.push_back(stoi(numbers.substr(0, next_pos)));
			numbers = numbers.substr(next_pos + 1, string::npos);
		}

		i_vect.push_back(stoi(numbers));

		// Adding the vector of numbers to their overall container
		series.push_back(i_vect);
	}

	// Looping over all the sequences
	for(int i = 0; i < sequences.size(); i++) {

		result = test_sequence(sequences[i], series[i]);
		solution_a += result;
		
		// Now computing the solution for part B by duplicating the sequences
		string duplicated_seq = sequences[i] + string("?") +
			sequences[i] + string("?") +
			sequences[i] + string("?") +
			sequences[i] + string("?") +
			sequences[i];

		vector<int> duplicated_series(0);

		for(int j = 0; j < 5; j++) duplicated_series.insert(duplicated_series.end(), series[i].begin(), series[i].end());

		result = test_sequence(duplicated_seq, duplicated_series);
		solution_b += result;
	}

	cout << "Puzzle 12A: " << solution_a << endl;
	cout << "Puzzle 12B: " << solution_b << endl;

	return 0;
}

// A function that preprocesses a sequence to determine the maximum
// number of successive hashes that can fit at a given position
vector<int> preprocess_sequence(const string& sequence) {
	vector<int> output (sequence.size());

	for(int i = 0; i < sequence.size(); i++) {
		int j = i;

		while(sequence[j] != '.' && j < sequence.size()) {
			output[i]++;
			j++;
		}
	}

	return output;
}

// A function that returns a boolean vector with the positions of hashes
vector<bool> find_hashes(const string& sequence) {
	vector<bool> output (sequence.size() + 1);

	for(int i = 0; i < sequence.size(); i++) {
		output[i] = (sequence[i] == '#');
	}

	output[sequence.size()] = false;

	return output;
}

// A function that tests a sequence and series combination
unsigned long long int test_sequence(const string& sequence, const vector<int>& series) {

	// Initializing the return value to 0
	unsigned long long int return_value = 0;

	// We first preprocess the sequence to identify the streaks of #? and the positions of the hash symbols
	vector<int> pp_vect = preprocess_sequence(sequence);
	vector<bool> h_vect = find_hashes(sequence);

	// Initializing a cache of the results to avoid computing them every time
	// The cache is a matrix with as many rows as sequence elements
	// and as many columns as series elements
	// It stores the number of possibilities associated with a given
	// combination of sequence position and series position
	// We also need a boolean matrix indicating whether the value is cached
	vector<vector<unsigned long long int>> cache_num(sequence.size());
	vector<vector<bool>> cache_bool(sequence.size());

	for(int i = 0; i < cache_num.size(); i++) {
		cache_num[i].resize(series.size());
		cache_bool[i].resize(series.size());
	}

	// Then we launch the recursive function with the indices that fit for the first series element
	int i = 0;

	while(i < sequence.size()) {
		if(pp_vect[i] >= series[0] && !h_vect[i + series[0]]) return_value += seq_recurse(sequence, pp_vect, h_vect, i, series, 0, cache_num, cache_bool);
		if(h_vect[i]) break;
		i++;
	}

	return return_value;
}

// A recursive function that finds the number of matches for a starting index in the sequence and index in the series of integers
unsigned long long int seq_recurse(const string& sequence, const vector<int>& pp_vect, const vector<bool>& h_vect,
		int index, const vector<int>& series, int series_index, vector<vector<unsigned long long int>>& cache_num, vector<vector<bool>>& cache_bool) {

	// If the value is cached then we return it
	if(cache_bool[index][series_index]) return cache_num[index][series_index];

	// Initializing the return value
	unsigned long long int return_value = 0;

	// We store the values of the indices that the function was called with
	int initial_index = index;
	int initial_sindex = series_index;

	// If this is the last element of the sequence then we can increment the return value
	// if there are no remaining hashes
	if(series_index == (series.size() - 1)) {
		index += series[series_index];

		while(index < sequence.size()) {
			if(h_vect[index]) {
				// We update the cache
				cache_bool[initial_index][initial_sindex] = true;
				cache_num[initial_index][initial_sindex] = 0;
				return 0;
			}
			index++;
		}

		// If there were no hashes remaining then this is a match and we increment the return value
		return_value++;
	} else {
		// If this is not the last series element, then we need to loop over possible indices
		// for the next element in the sequence

		// But first we increment the index and the series_index
		index += series[series_index] + 1;
		series_index++;

		// We cut the computation short if the number of elements to fit is too large compared to the remaining sequence
		// We initialize to -1 because the last series element does not need a space in between
		int to_fit = -1;
		for(int i = series_index; i < series.size(); i++) to_fit += series[i] + 1;

		// We go forward with the next element in the series if:
		// - there is enough room to fit the series element
		// - the element after the selected sequence is not a #
		// This loop is performed as long as the remaining sequence can fit
		// We also cannot leave a hash behind because that would break the requirements
		while(index < sequence.size() && to_fit <= (sequence.size() - index)) {
			if(pp_vect[index] >= series[series_index] && !h_vect[index + series[series_index]])
				return_value += seq_recurse(sequence, pp_vect, h_vect, index, series, series_index, cache_num, cache_bool);
			if(h_vect[index]) break;
			index++;
		}
	}

	// We update the cache prior to returning the result
	cache_bool[initial_index][initial_sindex] = true;
	cache_num[initial_index][initial_sindex] = return_value;
	return return_value;
}

